A36-xds-for-servers.md

```diff
@@ -0,0 +1,401 @@
+templates. If this becomes a problem more methods could become `virtual`.
+
+
+### Wrapped Languages
+
+This section is a sketch to convey the "feel" of the API. But details may vary.
+
+C will expose an API for wrapped languages to utilize. We do not focus on that
+here and leave that as an implementation detail to be part of a future gRFC.
+
+Each wrapped language will need a custom "xds server" creation API. This will
+vary per-language. We use Python here just as an example of how it may look.
+
+Python could add a `grpc.xds_server(...)` that mirrors `grpc.server(...)`. Since
+ports are added after the server is created, the returned `grpc.Server` object
+may be an xds-aware `Server` to coordinate with the C API when `server.start()`
+is called. However, its implementation should be trivial as it could mostly
+delegate to a normal `server` instance.
+
+It would be possible to manage the lifetime of the XdsClient from the `server`,
+based on `server.stop()`, but given C++ lacks this avenue the C API probably
+will not need this notification.
+
+
+### Java
+
+Create an `XdsServerBuilder` that extends `ServerBuilder` and delegates to a
+"real" builder. The server returned from `builder.build()` would be an xDS-aware
+server delegating to a "real" server instance. The `XdsServerBuilder` would
+install an `AtomicReference<ServerInterceptor>`-backed interceptor in the
+built server and pass the listening port and interceptor reference to the
+xDS-aware server when constructed. Since interceptors cannot be removed once
+installed, the builder may only be used once; `build()` should throw if called a
+second time. To allow the `XdsServerBuilder` to create addition servers if
+necessary, mutation of the builder after `build()` should also throw.
+
+The xDS-aware server’s `start()` will create the `XdsClient` with the passed
+port and wait for initial configuration before delegating to the real `start()`.
+If the xDS bootstrap is missing it will throw an IOException within `start()`.
+The `XdsClient` will be shut down when the server is terminated (generally
+noticed via `shutdownNow()`/`awaitTermination()`). When new or updated
+configuration is received, it will create a new intercepter and update the
+interceptor reference.
+
+XdsServerBuilder will have an
+`xdsServingStatusListener(XdsServingStatusListener)` method.
+`XdsServingStatusListener` will be defined as:
+
+```java
+package io.grpc.xds;
+
+public interface XdsServingStatusListener {
+  void onServing();
+  void onNotServing(Throwable t);
+}
+```
+
+It is an interface instead of an abstract class as additional methods are not
+expected to be added before Java 8 language features are permitted in the code
+base. If this proves incorrect, an additional interface can be added.
+
+If not specified, a default implementation of `XdsServingStatusListener` will be
+used. It will log the exception and log calls to `onServing()` following a call
+to `onNotServing()` at WARNING level.
```

Comment:
Shouldn't the first `onServing()` call also log i.e. the first time the server goes into serving mode?

Reply:
No. The point was to only log when something is going wrong, or has been resolved such that that previous log is no longer correct. We still don't want to log during normal operation.